File: mac_scripts/functions/brew-desired-update/brew.mjs
import { spawn as defaultSpawn } from "node:child_process";
import { accessSync, constants, existsSync } from "node:fs";
import { join } from "node:path";

const OPT_BREW = "/opt/homebrew/bin/brew";
const INTEL_BREW = "/usr/local/bin/brew";

const defaultFs = {
  existsSync,
  accessSync,
  constants,
};

/**
 * @param {import("node:fs")} fs
 * @param {string} filePath
 */
function isExecutable(fs, filePath) {
  if (!fs.existsSync(filePath)) {
    return false;
  }
  try {
    fs.accessSync(filePath, fs.constants?.X_OK ?? constants.X_OK);
    return true;
  } catch {
    return false;
  }
}

/**
 * @param {{ env?: Record<string, string | undefined>, fs?: typeof defaultFs, pathEnv?: string }} [options]
 * @returns {string | null}
 */
export function resolveBrewBinary({ env = process.env, fs = defaultFs, pathEnv } = {}) {
  const pathValue = pathEnv ?? env.PATH ?? "";
  for (const dir of pathValue.split(":")) {
    if (!dir) {
      continue;
    }
    const candidate = join(dir, "brew");
    if (isExecutable(fs, candidate)) {
      return candidate;
    }
  }
  if (isExecutable(fs, OPT_BREW)) {
    return OPT_BREW;
  }
  if (isExecutable(fs, INTEL_BREW)) {
    return INTEL_BREW;
  }
  return null;
}

/**
 * @param {string[]} args
 * @returns {string}
 */
export function formatBrewCommand(args) {
  const quoted = args.map((arg) => {
    if (arg === "") return "''";
    if (/^[A-Za-z0-9_./:=+-]+$/.test(arg)) return arg;
    return `'${String(arg).replace(/'/g, `'\\''`)}'`;
  });
  return `$ brew ${quoted.join(" ")}`;
}

/**
 * @param {string[]} args
 * @param {{
 *   brewBin: string,
 *   spawn?: typeof defaultSpawn,
 *   cwd?: string,
 *   onCommand?: (line: string) => void,
 *   stdout?: { write: (chunk: string | Buffer) => void },
 *   stderr?: { write: (chunk: string | Buffer) => void },
 * }} options
 * @returns {Promise<{ code: number, stdout: string, stderr: string }>}
 */
export function runBrew(
  args,
  {
    brewBin,
    spawn = defaultSpawn,
    cwd,
    onCommand = (line) => process.stdout.write(`${line}\n`),
    stdout: outStream = process.stdout,
    stderr: errStream = process.stderr,
  } = {},
) {
  return new Promise((resolve) => {
    onCommand(formatBrewCommand(args));
    const child = spawn(brewBin, args, {
      cwd,
      stdio: ["ignore", "pipe", "pipe"],
    });
    let stdout = "";
    let stderr = "";
    child.stdout?.on("data", (chunk) => {
      stdout += chunk;
      outStream.write(chunk);
    });
    child.stderr?.on("data", (chunk) => {
      stderr += chunk;
      errStream.write(chunk);
    });
    child.on("close", (code) => {
      resolve({ code: code ?? 1, stdout, stderr });
    });
    child.on("error", () => {
      resolve({ code: 1, stdout, stderr });
    });
  });
}

/**
 * @param {{
 *   brewBin: string,
 *   ui?: { command?: (line: string) => void },
 *   stdout?: { write: (chunk: string | Buffer) => void },
 *   stderr?: { write: (chunk: string | Buffer) => void },
 * }} options
 * @returns {(args: string[]) => Promise<{ code: number, stdout: string, stderr: string }>}
 */
export function createBrewRunner({
  brewBin,
  ui,
  stdout = process.stdout,
  stderr = process.stderr,
}) {
  return (args) => runBrew(args, {
    brewBin,
    stdout,
    stderr,
    onCommand: (line) => {
      if (ui?.command) ui.command(line);
      else stdout.write(`${line}\n`);
    },
  });
}

/**
 * @param {string} text
 * @returns {string[]}
 */
function parseLines(text) {
  return text.split(/\r?\n/).filter((line) => line.length > 0);
}

/**
 * @param {string} name
 * @returns {string}
 */
function shortName(name) {
  const slash = name.lastIndexOf("/");
  return slash === -1 ? name : name.slice(slash + 1);
}

/**
 * @param {{ brewBin: string, runBrew?: (args: string[]) => Promise<{ code: number, stdout: string, stderr: string }> }} options
 */
export async function loadBrewState({ brewBin, runBrew }) {
  const runner = runBrew ?? ((args) => runBrewImpl(args, { brewBin }));

  const [formulasResult, casksResult, tapsResult] = await Promise.all([
    runner(["list", "--formula"]),
    runner(["list", "--cask", "--full-name"]),
    runner(["tap"]),
  ]);

  return {
    formulas: parseLines(formulasResult.stdout).map(shortName),
    casks: parseLines(casksResult.stdout).map(shortName),
    taps: parseLines(tapsResult.stdout),
  };
}

/**
 * @param {"formula" | "cask"} type
 * @param {string} name
 * @param {{ brewBin: string, runBrew?: (args: string[]) => Promise<{ code: number, stdout: string, stderr: string }> }} deps
 * @returns {Promise<string | null>}
 */
export async function brewInfoToken(type, name, { brewBin, runBrew } = {}) {
  const runner = runBrew ?? ((args) => runBrewImpl(args, { brewBin }));
  const flag = type === "cask" ? "--cask" : "--formula";
  const { code, stdout } = await runner(["info", flag, name]);
  if (code !== 0) {
    return null;
  }
  const firstLine = stdout.split(/\r?\n/)[0] ?? "";
  const match = firstLine.match(/^==>\s+([^: (]+)/);
  return match ? match[1] : null;
}

/**
 * @param {string} tap
 * @param {{ brewBin: string, runBrew?: (args: string[]) => Promise<{ code: number, stdout: string, stderr: string }> }} deps
 * @returns {Promise<boolean>}
 */
export async function ensureTap(tap, { brewBin, runBrew } = {}) {
  const runner = runBrew ?? ((args) => runBrewImpl(args, { brewBin }));

  const trustHelp = await runner(["trust", "--help"]);
  if (trustHelp.code === 0) {
    const trust = await runner(["trust", "--tap", tap]);
    if (trust.code !== 0) {
      return false;
    }
  }

  const tapResult = await runner(["tap", tap]);
  return tapResult.code === 0;
}

/**
 * @param {{ brewBin: string, runBrew?: (args: string[]) => Promise<{ code: number, stdout: string, stderr: string }> }} deps
 * @returns {Promise<string[]>}
 */
export async function listBrewTaps({ brewBin, runBrew } = {}) {
  const runner = runBrew ?? ((args) => runBrewImpl(args, { brewBin }));
  const { code, stdout } = await runner(["tap"]);
  if (code !== 0) {
    return [];
  }
  return parseLines(stdout);
}

/**
 * @param {string[]} args
 * @param {{
 *   brewBin: string,
 *   onCommand?: (line: string) => void,
 *   stdout?: { write: (chunk: string | Buffer) => void },
 *   stderr?: { write: (chunk: string | Buffer) => void },
 * }} options
 */
function runBrewImpl(args, { brewBin, onCommand, stdout, stderr }) {
  return runBrew(args, { brewBin, onCommand, stdout, stderr });
}

File: mac_scripts/functions/brew-desired-update/brew.py
import os
import re
import subprocess
from concurrent.futures import ThreadPoolExecutor

OPT_BREW = "/opt/homebrew/bin/brew"
INTEL_BREW = "/usr/local/bin/brew"


def _is_executable(path: str) -> bool:
    return os.path.isfile(path) and os.access(path, os.X_OK)


def resolve_brew_binary(env: dict) -> str | None:
    path_value = env.get("PATH", "")
    for directory in path_value.split(":"):
        if not directory:
            continue
        candidate = os.path.join(directory, "brew")
        if _is_executable(candidate):
            return candidate
    if _is_executable(OPT_BREW):
        return OPT_BREW
    if _is_executable(INTEL_BREW):
        return INTEL_BREW
    return None


def format_brew_command(args: list[str]) -> str:
    quoted = []
    for arg in args:
        if arg == "":
            quoted.append("''")
        elif re.fullmatch(r"[A-Za-z0-9_./:=+-]+", arg):
            quoted.append(arg)
        else:
            quoted.append("'{}'".format(arg.replace("'", "'\\''")))
    return f"$ brew {' '.join(quoted)}"


def is_probe(args: list[str]) -> bool:
    if not args:
        return False
    cmd = args[0]
    rest = args[1:]
    if cmd in ("list", "info", "search"):
        return True
    if cmd == "tap" and not rest:
        return True
    if cmd == "trust" and "--help" in rest:
        return True
    if cmd == "--help":
        return True
    return False


class _NoOpUi:
    def command(self, line: str) -> None:
        pass


class BrewRunner:
    def __init__(self, brew_bin: str) -> None:
        self.brew_bin = brew_bin

    def run(self, args: list[str], ui) -> subprocess.CompletedProcess:
        ui.command(format_brew_command(args))
        command = [self.brew_bin, *args]
        if is_probe(args):
            return subprocess.run(
                command,
                stdin=subprocess.DEVNULL,
                capture_output=True,
                text=True,
                check=False,
            )
        result = subprocess.run(command, check=False)
        return subprocess.CompletedProcess(
            args=command,
            returncode=result.returncode,
            stdout="",
            stderr="",
        )


def _parse_lines(text: str) -> list[str]:
    return [line for line in re.split(r"\r?\n", text) if line]


def _short_name(name: str) -> str:
    slash = name.rfind("/")
    return name[slash + 1:] if slash != -1 else name


def load_brew_state(runner: BrewRunner, ui=None) -> dict:
    log_ui = ui if ui is not None else _NoOpUi()

    def run_probe(args: list[str]) -> subprocess.CompletedProcess:
        return runner.run(args, log_ui)

    with ThreadPoolExecutor(max_workers=3) as pool:
        formulas_future = pool.submit(run_probe, ["list", "--formula"])
        casks_future = pool.submit(run_probe, ["list", "--cask", "--full-name"])
        taps_future = pool.submit(run_probe, ["tap"])
        formulas_result = formulas_future.result()
        casks_result = casks_future.result()
        taps_result = taps_future.result()

    return {
        "formulas": [_short_name(n) for n in _parse_lines(formulas_result.stdout)],
        "casks": [_short_name(n) for n in _parse_lines(casks_result.stdout)],
        "taps": _parse_lines(taps_result.stdout),
    }


def ensure_tap(tap: str, runner: BrewRunner) -> bool:
    silent = _NoOpUi()
    trust_help = runner.run(["trust", "--help"], silent)
    if trust_help.returncode == 0:
        trust = runner.run(["trust", "--tap", tap], silent)
        if trust.returncode != 0:
            return False
    tap_result = runner.run(["tap", tap], silent)
    return tap_result.returncode == 0
